src/ssl/gadgets.cc

```diff
@@ -15,6 +15,48 @@
 #include "security/Io.h"
 #include "ssl/gadgets.h"
 
+// Determines whether the given key requires a digest when signing.
+static bool
+keyNeedsDigest(const EVP_PKEY * const pkey) {
+    const auto id = EVP_PKEY_id(pkey);
+
+    // For provider-based keys, id is -1
+    if (id == -1) {
+        // Try all known ML-DSA names
+        if (EVP_PKEY_is_a(pkey, "ML-DSA") ||
+            EVP_PKEY_is_a(pkey, "ML-DSA-44") ||
+            EVP_PKEY_is_a(pkey, "ML-DSA-65") ||
+            EVP_PKEY_is_a(pkey, "ML-DSA-87"))
+            return false;
+    }
+
+    switch (id) {
+        case EVP_PKEY_RSA:
+        case EVP_PKEY_RSA_PSS:
+        case EVP_PKEY_DSA:
+        case EVP_PKEY_EC:
+            return true;  // classic algorithms need digest like SHA256
+
+        case EVP_PKEY_ED25519:
+        case EVP_PKEY_ED448:
+#ifdef EVP_PKEY_ML_DSA_44
+        case EVP_PKEY_ML_DSA_44: // post-quantum ML-DSAs
+#endif
+#ifdef EVP_PKEY_ML_DSA_65
+        case EVP_PKEY_ML_DSA_65:
+#endif
+#ifdef EVP_PKEY_ML_DSA_87
+        case EVP_PKEY_ML_DSA_87:
+#endif
+            return false; // digest must be NULL
+
+        default:
+            // Unknown type — safest is to assume digest required
+            // For provider-based keys, id is -1
+            return true;
+    }
+}
+
```

Comment:
```suggestion
keyNeedsDigest(const EVP_PKEY * const pkey) {
    if (EVP_PKEY_is_a(pkey, "ML-DSA") ||
        EVP_PKEY_is_a(pkey, "ML-DSA-44") ||
        EVP_PKEY_is_a(pkey, "ML-DSA-65") ||
        EVP_PKEY_is_a(pkey, "ML-DSA-87") ||
        EVP_PKEY_is_a(pkey, "ED25519") ||
        EVP_PKEY_is_a(pkey, "ED448"))
        return false; // no digest needed

    return true; // require a digest for all other type
}

```

I think it should be enough to rewrite the function like this, since I learned that provider-only algorithms (such as ML-DSA) do not have a legacy NID/ID. The EVP_PKEY_id/EVP_PKEY_base_id API should not be used; instead, EVP_PKEY_get0_type_name() or EVP_PKEY_is_a() should be used.

Would you @rousskov agree with the proposed change? I also tested Squid with ED25519/ED448 pkey types, and Squid failed in the same way when using the original code. With the changes in the original commit, or even with these suggested changes, Squid no longer refuses to start with the following error:

```
2025/09/03 06:22:11| FATAL: Unable to generate signing certificate for untrusted sites for HTTPS_port [::]:3128
```

Reply:
I have finally resumed working on this PR and am making progress. Should be done by Monday.